migrator/migrator.go

```diff
@@ -0,0 +1,97 @@
+package migrator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/microstorage"
+)
+
+const (
+	repoSlug = "github.com/giantswarm/microstorage"
+
+	migrationKey = repoSlug + "/migration"
+	migrationVal = "v1"
+)
+
+type Config struct {
+	Logger micrologger.Logger
+}
+
+// DefaultConfig creates a new configuration with the default settings.
+func DefaultConfig() Config {
+	return Config{
+		Logger: nil, // Required.
+	}
+}
+
+type Migrator struct {
+	logger micrologger.Logger
+}
+
+func New(config Config) (*Migrator, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger is empty")
+	}
+
+	m := &Migrator{
+		logger: config.Logger,
+	}
+
+	return m, nil
+}
+
+func (m *Migrator) Migrate(ctx context.Context, dst, src microstorage.Storage) error {
+	var err error
+
+	m.logger.Log("debug", "checking if migraiton is already done")
+	{
+		v, err := dst.Search(ctx, migrationKey)
+		if microstorage.IsNotFound(err) {
+			// This means migration hasn't been done yet.
+		} else if err != nil {
+			return microerror.Maskf(err, "dst storage: searching key=%s", migrationKey)
+		} else if v == migrationVal {
+			m.logger.Log("info", "migration already done")
+			return nil
+		}
+	}
```

Comment:
The migration logic uses additional state to reason about the migration itself. IMO the state the migration wants to migrate is the source of truth and only authority to reason about the state of the migration. I see this from the operator point of view. Operators do not create additional state to ensure that resources do not have to be reconciled anymore. Resources always have to be reconciled, otherwise the reconciliation is meaningless. The additional state hijacks the overarching goal of reconciling from the current state to the desired state. So IMO the migration should go through all the data and migrate each key if it is not yet migrated.